### ISSUE TITLE
High memory usage when sending to CloudWatch Logs on 0.21.0

### ISSUE THREAD

**Issue by wtchangdm** [vectordotdev/vector#12246]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Hi, we did a stress test again after 0.21.0 was released eariler today.

We noticed that when trying to send logs to CloudWatch Logs from Kubernetes Logs (not sure whether source matters or not here), the memory was significantly higher than it was in 0.20.1.

To compare the difference between these two versions, I ran the log generator pod (manifest was attached below) with vector `0.20.1-distroless-static` and then `0.21.0-distroless-static`.

### Resource limit for vector container (Pod deployed as daemonset)
```yaml
resources:
  requests:
    cpu: 200m
    memory: 1Gi
  limits:
    cpu: 500m
    memory: 1Gi
```

### Container status

![](https://i.imgur.com/ueOPZ3u.png)
([full size image](https://i.imgur.com/ueOPZ3u.png))

### Test with `0.20.1-distroless-static`
- Log generator was running from 21:58 to 22:08. Then manually scaled replicas from 1 to 0.
- CPU was stuck to 100% (its burst limit `500m`)
- Memory usage started at 11MiB at the very beginning and slowly increased to 150MiB.
- Container never exited.

### Test with `0.21.0-distroless-static`
- Log generator was running from 22:31 to 22:41. Then manually scaled replicas from 1 to 0.
- CPU was less than 100%
- Memory usage started at 12MiB  at the very beginning and suddenly burst to 492MiB, while the container was restarted a few times.
- The container itself was not OOMKilled. And weirdly the exit code was `0`. The following are the last logs before exited.

    ```
    2022-04-15T14:42:57.071188Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="cloudwatch_logs" time_remaining="49 seconds left"
    ...
    2022-04-15T14:42:58.161659Z TRACE sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}: vector_core::stream::driver: Acking events. ack_size=712
    2022-04-15T14:42:58.161672Z DEBUG sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}: vector::topology::builder: Finished.
    2022-04-15T14:42:58.161707Z DEBUG vector_buffers::variants::disk_v1::reader: Running flush.
    2022-04-15T14:42:58.161715Z TRACE vector_buffers::topology::acks: Added acknowledgements. unclaimed_acks=15724 added_acks=15310
    2022-04-15T14:42:58.161721Z TRACE vector_buffers::topology::acks: Searching for eligible marker. acked_marker_id=1379499 effective_acked_marker_id=1395223 unclaimed_acks=15724 pending_markers=26
    ...
    2022-04-15T14:42:58.161833Z DEBUG vector_buffers::variants::disk_v1::reader: Deleting 26 records from buffer: 15724 items, 20672013 bytes. delete_offset=1395223
    2022-04-15T14:42:58.161869Z DEBUG vector_buffers::variants::disk_v1::reader: Compacting disk buffer.
    ```
- Then the container exited and restarted.

I made a very simple log generator to reproduce this issue. Please check the following collapsed block.
<details>
<summary>Click to expand log generator YAML</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: log-test
  name: log-test
  namespace: default
spec:
  replicas: 0
  selector:
    matchLabels:
      app: log-test
  template:
    metadata:
      labels:
        app: log-test
    spec:
      containers:
      - image: wtchangdm/log-test:0.0.1
        name: printlog
        resources:
          limits:
            cpu: "1"
            memory: 100Mi
          requests:
            cpu: "1"
            memory: 100Mi
```
</details>

### Configuration

```text
# Although it contains the s3 sink, we only need to trigger the cloudwatch one to reproduce. 

data_dir = "/vector-data-dir"

[api]
enabled = true
address = "0.0.0.0:8686"

[sources.internal_metrics]
type = "internal_metrics"

scrape_interval_secs = 2


[sources.k8s_logs]
type                     = "kubernetes_logs"
delay_deletion_ms        = 60000
glob_minimum_cooldown_ms = 5000

extra_field_selector        = "metadata.namespace=default"

exclude_paths_glob_patterns = ["**/*.gz","**/*.tmp","/var/log/pods/**/*vector*/**"]

[sources.k8s_logs.pod_annotation_fields]
container_name = "kube_container_name"
pod_namespace  = "kube_namespace"
pod_name       = "source"


[transforms.sanitize_k8s_logs]
type = "remap"
inputs = ["k8s_logs"]

reroute_dropped = true
drop_on_abort = true
drop_on_error = true

source = """
  result, err = parse_json(.message)

  if err != null {
    result = {
      "log": .message
    }
  }

  result.kube_namespace = .kube_namespace
  result.kube_container_name = .kube_container_name
  result.source = .source
  result.stream = .stream
  result.timestamp = .timestamp

  . = result
"""


[transforms.filter_for_s3]
type = "filter"
inputs = ["sanitize_k8s_logs"]

condition = 'includes(["debug","<REDACTED>"], .logType)'


[transforms.filter_for_cloudwatch]
type = "filter"
inputs = ["sanitize_k8s_logs"]

condition = '!exists(.logType) || includes(["audit"], .logType)'


[sinks.cloudwatch_logs]
inputs = ["filter_for_cloudwatch"]
type = "aws_cloudwatch_logs"

# Note: dynamic log group cannot be examined at start.
healthcheck.enabled = false

region = "ap-northeast-1"
create_missing_group = true
create_missing_stream = true
group_name = "/<REDACTED_PREFIX>/{{ kube_namespace }}"
compression = "gzip"

stream_name = "{{ source }}-{{ kube_container_name }}-via-${VECTOR_SELF_NODE_NAME}"
encoding = "json"

buffer.type = "disk"
buffer.max_size = 150000000
buffer.when_full = "block"


[sinks.s3]
inputs = ["filter_for_s3"]
type = "aws_s3"

healthcheck.enabled = true

region = "ap-northeast-1"
bucket = "<REDACTED>"
key_prefix = "<REDACTED_PREFIX>-{{ kube_namespace }}/%Y/%m/%d/%H/"
filename_append_uuid = true
compression = "gzip"

batch.timeout_secs = 300
batch.max_bytes = 10000000

buffer.type = "disk"
buffer.max_size = 150000000
buffer.when_full = "block"

encoding.codec = "ndjson"


[sinks.cloudwatch_logs_for_dropped_records]
inputs = ["sanitize_k8s_logs.dropped"]
type = "aws_cloudwatch_logs"

region = "ap-northeast-1"
create_missing_group = true
create_missing_stream = true
compression = "gzip"
healthcheck.enabled = false
group_name = "/<REDACTED_PREFIX>/vector-dropped-records"
stream_name = "{{ source }}-{{ kube_container_name }}-from-${VECTOR_SELF_NODE_NAME}"
encoding = "json"

buffer.type = "disk"
buffer.max_size = 150000000
buffer.when_full = "block"


[sinks.prometheus_exporter]
type = "prometheus_exporter"
inputs = ["internal_metrics"]

address = "0.0.0.0:9090"
default_namespace = "vector"
```

### Version

vector 0.21.0 (x86_64-unknown-linux-musl c1edb89 2022-04-14)

### Debug Output

```text
The following are part of the logs from 0.21.0 before it exited for an unknown reason.

The complete file is ~200MB before compression. If it's needed, I will find some time to review the sensitive info and then send it to you.


2022-04-15T14:30:41.265752Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}: vector::internal_events::common: Bytes received. byte_size=1191 protocol=http
2022-04-15T14:30:41.265790Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}:file_server: file_source::file_server: Read bytes. path="/var/log/pods/default_log-test-7b48dc94cc-5wmwc_8617908c-c014-41f8-ae06-666dcfcfe106/printlog/0.log" bytes=1191
2022-04-15T14:30:41.265834Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}:file_server: file_source::file_server: Read bytes. path="/var/log/pods/default_log-test-7b48dc94cc-5wmwc_8617908c-c014-41f8-ae06-666dcfcfe106/printlog/0.log" bytes=1191
2022-04-15T14:30:41.265858Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}: vector::internal_events::kubernetes_logs: Events received. count=1 byte_size=3545 file=/var/log/pods/default_log-test-7b48dc94cc-5wmwc_8617908c-c014-41f8-ae06-666dcfcfe106/printlog/0.log
2022-04-15T14:30:41.266119Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}: vector::internal_events::common: Bytes received. byte_size=1191 protocol=http
2022-04-15T14:30:41.266195Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}: vector::internal_events::kubernetes_logs: Events received. count=1 byte_size=3545 file=/var/log/pods/default_log-test-7b48dc94cc-5wmwc_8617908c-c014-41f8-ae06-666dcfcfe106/printlog/0.log
2022-04-15T14:30:41.266432Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}: vector::internal_events::common: Bytes received. byte_size=1191 protocol=http
2022-04-15T14:30:41.266447Z TRACE source{component_kind="source" component_id=k8s_logs component_type=kubernetes_logs component_name=k8s_logs}:file_server: file_source::file_server: Read bytes. path="/var/log/pods/default_log-test-7b48dc94cc-5wmwc_8617908c-c014-41f8-ae06-666dcfcfe106/printlog/0.log" bytes=1191
# ...
```

The only errors were about rate exceeded of certain CloudWatch Log Stream:

```
2022-04-15T14:31:39.190381Z ERROR sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}:request{request_id=115}: vector::sinks::util::retries: Non-retriable error; dropping the request. error=CloudwatchError::Put: Error { code: "ThrottlingException", message: "Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal", request_id: "e7f5711a-665b-4714-90ae-76d287dabdxxx" }
2022-04-15T14:31:39.190625Z ERROR sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}:request{request_id=115}: vector::sinks::util::retries: Non-retriable error; dropping the request. error=CloudwatchError::Put: Error { code: "ThrottlingException", message: "Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal", request_id: "e7f5711a-665b-4714-90ae-76d287dabdxxx" }
2022-04-15T14:31:39.190721Z ERROR sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}:request{request_id=115}: vector_core::stream::driver: Service call failed. error=Put(ServiceError { err: PutLogEventsError { kind: Unhandled(Error { code: Some("ThrottlingException"), message: Some("Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal"), request_id: Some("e7f5711a-665b-4714-90ae-76d287dabdxxx"), extras: {} }), meta: Error { code: Some("ThrottlingException"), message: Some("Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal"), request_id: Some("e7f5711a-665b-4714-90ae-76d287dabdxxx"), extras: {} } }, raw: Response { inner: Response { status: 400, version: HTTP/1.1, headers: {"x-amzn-requestid": "e7f5711a-665b-4714-90ae-76d287dabdxxx", "content-type": "application/x-amz-json-1.1", "content-length": "162", "date": "Fri, 15 Apr 2022 14:31:38 GMT", "connection": "close"}, body: SdkBody { inner: Once(Some(b"{\"__type\":\"ThrottlingException\",\"message\":\"Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal\"}")), retryable: true } }, properties: SharedPropertyBag(Mutex { data: PropertyBag, poisoned: false, .. }) } }) request_id=115
```

### Example Data

```json
{"level":30,"time":1650036052106,"pid":8,"hostname":"7d5022dfea8a","logType":"audit","message":"longtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtextlongtext"}
```

(~1.1KB)

### Additional Context

I am not sure whether it's related to the AWS SDK change in `0.21.0`. But I think it should be easy to reproduce.

Please let me know if there is any information I can provide to help track down this issue.

Thanks.

P.S. Here is a little off-topic question. I saw the error from CloudWatch Logs that said it exceeded the rate limit. The error looked like this: 

```
2022-04-15T14:31:39.190625Z ERROR sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}:request{request_id=115}: vector::sinks::util::retries: Non-retriable error; dropping the request. error=CloudwatchError::Put: Error { code: "ThrottlingException", message: "Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal", request_id: "e7f5711a-665b-4714-90ae-76d287dabdxxx" }
2022-04-15T14:31:39.190721Z ERROR sink{component_kind="sink" component_id=cloudwatch_logs component_type=aws_cloudwatch_logs component_name=cloudwatch_logs}:request{request_id=115}: vector_core::stream::driver: Service call failed. error=Put(ServiceError { err: PutLogEventsError { kind: Unhandled(Error { code: Some("ThrottlingException"), message: Some("Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal"), request_id: Some("e7f5711a-665b-4714-90ae-76d287dabdxxx"), extras: {} }), meta: Error { code: Some("ThrottlingException"), message: Some("Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal"), request_id: Some("e7f5711a-665b-4714-90ae-76d287dabdxxx"), extras: {} } }, raw: Response { inner: Response { status: 400, version: HTTP/1.1, headers: {"x-amzn-requestid": "e7f5711a-665b-4714-90ae-76d287dabdxxx", "content-type": "application/x-amz-json-1.1", "content-length": "162", "date": "Fri, 15 Apr 2022 14:31:38 GMT", "connection": "close"}, body: SdkBody { inner: Once(Some(b"{\"__type\":\"ThrottlingException\",\"message\":\"Rate exceeded for logStreamName log-test-7b48dc94cc-5wmwc-printlog-via-ip-xx-xx-xx-xx.ap-northeast-1.compute.internal\"}")), retryable: true } }, properties: SharedPropertyBag(Mutex { data: PropertyBag, poisoned: false, .. }) } }) request_id=115
```

Why is this `Non-retriable error`? I thought it's [retryable][1] (and there is `retryable: true` in the log). I can open another issue about this if needed.

[1]: https://aws.amazon.com/premiumsupport/knowledge-center/cloudwatch-logs-throttling/

**Comment by jszwedko** [comment]
Thanks for the detailed issue report and reproduction case (including the log generator and manifests)! We will look into this.

**Comment by wtchangdm** [comment]
Yeah, through helm chart. But it's a in-house one that we originally used
for some internal requirements with fluentbit, and I modified it in order
to switch to vector.

Is there any problem when reproducing this issue?

**Comment by jszwedko** [comment]
> Is there any problem when reproducing this issue?

I actually am having some trouble reproducing this 😓 

I was able to run a similar setup to yours:

Log generation:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: log-test
  name: log-test
  namespace: default
spec:
  replicas: 0
  selector:
    matchLabels:
      app: log-test
  template:
    metadata:
      labels:
        app: log-test
    spec:
      containers:
      # I tried your log generation image above but it just exited immediately for some reason
      - image: mingrammer/flog:latest
        name: printlog
        command: ["flog", "-f", "json", "-l"]
        resources:
          limits:
            cpu: "1"
            memory: 100Mi
          requests:
            cpu: "1"
            memory: 100Mi
```

Deploying Vector via the helm chart with `values.yaml` (I used your config minus the S3 bits since you said you could reproduce without them):

```yaml
# role -- Role for this Vector (possible values: Agent, Aggregator, Stateless-Aggregator)
## Ref: https://vector.dev/docs/setup/deployment/roles/
## Each role is created with the following workloads:
## Agent - DaemonSet
## Aggregator - StatefulSet
## Stateless-Aggregator - Deployment
role: "Agent"

## Define the Vector image to use
image:
  # image.repository -- Override default registry + name for Vector
  repository: timberio/vector
  # image.tag -- Vector image tag to use
  # @default -- Chart's appVersion
  #tag: "0.20.1-distroless-static"
  tag: "0.21.0-distroless-static"

## Create a Secret resource for Vector to use
secrets:
  # secrets.generic -- Each Key/Value will be added to the Secret's data key, each value should be raw and NOT base64 encoded
  ## Any secrets can be provided here, it's commonly used for credentials and other access related values.
  ## NOTE: Don't commit unencrypted secrets to git!
  generic:
    awsAccessKeyId: <REDACTED>
    awsSecretAccessKey: <REDACTED>
    awsSessionToken: <REDACTED>

# env -- Set environment variables in Vector containers
## The examples below leverage examples from secrets.generic and assume no name overrides with a Release name of "vector"
env:
  - name: AWS_ACCESS_KEY_ID
    valueFrom:
      secretKeyRef:
        name: vector
        key: awsAccessKeyId
  - name: AWS_SECRET_ACCESS_KEY
    valueFrom:
      secretKeyRef:
        name: vector
        key: awsSecretAccessKey
  - name: AWS_SESSION_TOKEN
    valueFrom:
      secretKeyRef:
        name: vector
        key: awsSessionToken

# resources -- Set Vector resource requests and limits.
#resources: {}
resources:
  requests:
    cpu: 200m
    memory: 1Gi
  limits:
    cpu: 500m
    memory: 1Gi

# customConfig -- Override Vector's default configs, if used **all** options need to be specified
## This section supports using helm templates to populate dynamic values
## Ref: https://vector.dev/docs/reference/configuration/
customConfig:
  data_dir: /vector-data-dir
  api:
    enabled: true
    address: '0.0.0.0:8686'
  sources:
    internal_metrics:
      type: internal_metrics
      scrape_interval_secs: 2
    k8s_logs:
      type: kubernetes_logs
      delay_deletion_ms: 60000
      glob_minimum_cooldown_ms: 5000
      extra_field_selector: metadata.namespace=default
      exclude_paths_glob_patterns:
        - '**/*.gz'
        - '**/*.tmp'
        - /var/log/pods/**/*vector*/**
      pod_annotation_fields:
        container_name: kube_container_name
        pod_namespace: kube_namespace
        pod_name: source
  transforms:
    sanitize_k8s_logs:
      type: remap
      inputs:
        - k8s_logs
      reroute_dropped: true
      drop_on_abort: true
      drop_on_error: true
      source: |
          result, err = parse_json(.message)

          if err != null {
            result = {
              "log": .message
            }
          }

          result.kube_namespace = .kube_namespace
          result.kube_container_name = .kube_container_name
          result.source = .source
          result.stream = .stream
          result.timestamp = .timestamp

          . = result
    filter_for_cloudwatch:
      type: filter
      inputs:
        - sanitize_k8s_logs
      condition: '!exists(.logType) || includes(["audit"], .logType)'

  sinks:
    cloudwatch_logs:
      inputs:
        - filter_for_cloudwatch
      type: aws_cloudwatch_logs
      healthcheck:
        enabled: false
      region: us-east-1
      create_missing_group: true
      create_missing_stream: true
      group_name: '/jesse-vector-test/{{ "{{" }} kube_namespace {{ "}}" }}'
      compression: gzip
      stream_name: '{{ "{{" }} source {{ "}}" }}-{{ "{{" }} kube_container_name {{ "}}" }}-via-me'
      encoding: json
      buffer:
        type: disk
        max_size: 150000000
        when_full: block
    cloudwatch_logs_for_dropped_records:
      inputs:
        - sanitize_k8s_logs.dropped
      type: aws_cloudwatch_logs
      region: us-east-1
      create_missing_group: true
      create_missing_stream: true
      compression: gzip
      healthcheck:
        enabled: false
      group_name: /jesse-vector-test/vector-dropped-records
      stream_name: '{{ "{{" }} source {{ "}}" }}-{{ "{{" }} kube_container_name {{ "}}" }}-from-me'
      encoding: json
      buffer:
        type: disk
        max_size: 150000000
        when_full: block
    prometheus_exporter:
      type: prometheus_exporter
      inputs:
        - internal_metrics
      address: '0.0.0.0:9090'
      default_namespace: vector
```

Running with `0.20.1-distroless-static` first I capped out at around 4 MiB of memory usage. Running with `0.20.1-distroless-static` I actually capped out around 2 MiB of memory usage:

<img width="1886" alt="Screen Shot 2022-04-21 at 11 02 01 AM" src="https://user-images.githubusercontent.com/316880/164522915-b02558a7-c087-409a-ac60-919c2241c60c.png">

(ignoring the first two, the 3rd one is 0.20.1 and the 4th one is 0.21.0)

I'm guessing there are other variables at play here. Some thoughts:

* Maybe your log generate outputs more quickly?
* I was running locally. I imagine you are running in AWS and so can achieve higher throughput pushing events to CloudWatch Logs

Could you share the `vector_buffer_events_sent_total`? That would help me compare throughput. I need to adjust my deployment to capture that.

**Comment by wtchangdm** [comment]
> I tried your log generation image above but it just exited immediately for some reason

> Maybe your logging pod outputs more quickly?

Sorry to hear that, it's just a process spamming the same logs over and over. I think it's more aggressive and has larger size per JSON line than the one you replaced with. (You may try `docker run --rm -it wtchangdm/log-test:0.0.1`)
I tried it in EKS and local k3d, and it works. 

Have you scaled the replica to 1? It was 0 in the manifest just to make sure logs are only generated when everything is set.
Otherwise, was there any status that showed up after `$ kubeectl get po -o yaml <log-test-pod>`?

> I was running locally. I imagine you are running in AWS and so can achieve higher throughput pushing events to CloudWatch Logs

Yes, I tested this in an EKS 1.21 Cluster with a self-managed node group. The instance was `c5.xlarge`.


> Could you share the vector_buffer_events_sent_total?

Yeah, I think I have attached it in the original post? It's about 1MiB/s in `0.20.1` and roughly 2~3 MiB/s in `0.21.0`.

I created that panel with the following PromQL:

```
sum(rate(vector_buffer_sent_bytes_total{component_id="cloudwatch_logs"}[2m])) by (pod, component_id)
```

The following is the helm-rendered DaemonSet Vector workload (config is the same as above). Hope it helps.

<details>
<summary>Click to expand log Vector DaemonSet</summary>

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: release-name
  namespace: cluster-log-shipper
  labels:
    k8s-app: release-name
    version: v1
    kubernetes.io/cluster-service: "true"
spec:
  selector:
    matchLabels:
      k8s-app: release-name
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 25%
  template:
    metadata:
      labels:
        k8s-app: release-name
        version: v1
        kubernetes.io/cluster-service: "true"
    spec:
      serviceAccountName: "<REDACTED>"
      containers:
        - name: vector
          image: "timberio/vector:0.20.1-distroless-libc"
          args:
            - -vvv
          ports:
            - name: prom-exporter
              containerPort: 9090
              protocol: TCP
            - name: api
              containerPort: 8686
              protocol: TCP
          readinessProbe:
            failureThreshold: 3
            httpGet:
              path: /health
              port: api
          livenessProbe:
            failureThreshold: 6
            httpGet:
              path: /health
              port: api
          env:
            - name: VECTOR_SELF_NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
            - name: VECTOR_SELF_POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: VECTOR_SELF_POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            # comment this in order to make -vvv work
            # - name: VECTOR_LOG
            #   value: "info"
            - name: PROCFS_ROOT
              value: "/host/proc"
            - name: SYSFS_ROOT
              value: "/host/sys"
          resources:
            limits:
              cpu: 500m
              memory: 250Mi
            requests:
              cpu: 200m
              memory: 250Mi
          volumeMounts:
            - name: config
              mountPath: /etc/vector/
              readOnly: true
            - name: data
              mountPath: /vector-data-dir
            - name: var-log
              mountPath: /var/log
              readOnly: true
            - name: var-lib
              mountPath: "/var/lib"
              readOnly: true
            - name: procfs
              mountPath: "/host/proc"
              readOnly: true
            - name: sysfs
              mountPath: "/host/sys"
              readOnly: true
      terminationGracePeriodSeconds: 30
      volumes:
        - name: config
          projected:
            sources:
              - configMap:
                  name: "release-name-config"
        - name: data
          hostPath:
            path: "/var/lib/vector"
        - name: var-log
          hostPath:
            path: "/var/log/"
        - name: var-lib
          hostPath:
            path: "/var/lib/"
        - name: procfs
          hostPath:
            path: "/proc"
        - name: sysfs
          hostPath:
            path: "/sys"
      tolerations:
      - operator: Exists
```

**Comment by jszwedko** [comment]
Hey @wtchangdm !

I took another pass at this today, setting up an EKS cluster in AWS with `c5.xlarge` for the node group and using your log generator. Unfortunately, I'm still not able to reproduce. I am getting roughly the same throughput (2-3 Mbps) but memory and CPU quickly stabilizes for both 0.20.1 and 0.21.0.

First run, 0.20.1:

<img width="1764" alt="Screen Shot 2022-05-03 at 4 09 01 PM" src="https://user-images.githubusercontent.com/316880/166590371-49736786-b472-4da8-be71-f62ce7c792e7.png">

You can see there a let log generation run for about 20 minutes. CPU use stabilized around 30% with memory around 6% (~60 MB).

Second run, 0.21.0:

<img width="1764" alt="Screen Shot 2022-05-03 at 4 39 11 PM" src="https://user-images.githubusercontent.com/316880/166591354-89324e3f-81bf-4b56-b62c-2a96c89c4657.png">

CPU usage was roughly similar. Memory use was a little higher, but not at the levels you were seeing (around 7-10% or 70-100 MB).

I feel like I'm probably missing something from your setup, but I'm not sure what it might be.

If you are able to consistently reproduce, we could try to bisect through together: starting by bisecting the nightly images. Would you be open to that? We'd want to narrow it down from the nightly images published from 2022-02-10 until 2022-04-07. These are tagged like: [timberio/vector:nightly-2022-05-03-distroless-static](https://hub.docker.com/layers/vector/timberio/vector/nightly-2022-05-03-distroless-static/images/sha256-c0212dc89654d8e38f24371787a9263109f376efc230145e7ae56a9753ff2222?context=explore) so you could try running the first one, last one, and then split in the middle until you find the day that results in the increase (hopefully the increase is centered on one day).

**Comment by wtchangdm** [comment]
@jszwedko Thank you for the effort and reply! Of course I am open to the suggestion as long as I can help on the issue.

I've tested with the nightly builds and was able to narrow it down a bit:

![img](https://i.imgur.com/05HUzEb.png)

From left to right:

- vector-6gmjs (0.20.1-distroless-static)
- vector-xkjng (nightly-2022-02-25-distroless-static)
- vector-2bqsd (nightly-2022-02-26-distroless-static)
- vector-gdwhl (0.21.0-distroless-static) (restarted once)

As you can see, `nightly-2022-02-26-distroless-static`'s memory usage is higher than `02-25`'s build. So:

- The last working nightly build image without memory spike: `nightly-2022-02-25-distroless-static` (Still higher than yours. Not sure why.)
- The first image that shows memory spikes under the same load: `nightly-2022-02-26-distroless-static`
- `0.21.0`'s memory usage is much higher than `nightly-2022-02-26-distroless-static`, plus restarting with exit code `0`.

Some thoughts on the differences:
1. Are you using [regions that have higher `PutLogEvents`][1] rate limit (e.g. Virginia, Oregon)? You can see it in Service Quota console. Mine is 800/s  without any adjustment. 
    - (although it looks a bit unlikely since #12255 shares the same symptom)
3. We use [default kubelet config][2] with `10Mi` for `containerLogMaxSize` and `5` for `containerLogMaxFiles`. Judging by the vector logs, container logs are rotated pretty frequently, not sure if yours are the same (if your runtime is Docker than there might be a difference).
4. I am using [IRSA][3] to grand IAM Role permission to Pod level. Not sure it made things different. If you use instance role or access key directly, I might need to control this variable as well.
5. The EKS cluster I tested is for QA purpose but with almost the same configuration as Prod. I can create a brand new cluster if needed. It's just weird that Vector behaved before I changed image version.

Some other info:
- Region: ap-northeast-1 (Tokyo)
- EKS control plane version: 
    ```
    Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.9-eks-0d102a7", GitCommit:"eb09fc479c1b2bfcc35c47416efb36f1b9052d58", GitTreeState:"clean", BuildDate:"2022-02-17T16:36:28Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
    ```
- Self-managed worker node with AMI: `amazon-eks-node-1.21-v20220420`
  - Container runtime: `containerd github.com/containerd/containerd 1.4.13 9cc61520f4cd876b86e77edfeb88fbcd536d1f9d`
- The worker node under testing doesn't have any other settings, not even extra labels or taints.


I was thinking maybe I can send my helm chart to you for the testing via something like email or discord? There is nothing secret, I just rather not share it publicly.

If there is anything I can do for help, please let me know.

Thanks.


[1]: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html
[2]: https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/
[3]: https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html

**Comment by jszwedko** [comment]
Thanks for narrowing it down! I appreciate you helping us troubleshoot this.

My suspicion looking at the [diff between `nightly-2022-02-25-distroless-static` and `nightly-2022-02-26-distroless-static`](https://github.com/vectordotdev/vector/compare/cc960fa...71d067f) is this commit to start sending events through Vector in batches: https://github.com/vectordotdev/vector/commit/6dd9d65c85368200434ef68ae76430e8f7dbe64e

I'd like to build an image of both that commit and the previous commit for you to confirm if you are open to running them.

To answer your other questions:

> * We use [default kubelet config](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) with `10Mi` for `containerLogMaxSize` and `5` for `containerLogMaxFiles`. Judging by the vector logs, container logs are rotated pretty frequently, not sure if yours are the same (if your runtime is Docker than there might be a difference).

I'm also using the default kubelet config. Kubernetes version 1.21. EKS platform version `eks.6`.

> * I am using [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) to grand IAM Role permission to Pod level. Not sure it made things different. I might need to control this variable as well.

Also using IRSA.

If we can narrow it down to that commit, hopefully we can figure out possible areas to investigate 🤞 .

**Comment by jszwedko** [comment]
Hi @wtchangdm !

I pushed up two images for you to try:

jszwedkodd/vector:b8fb1e388-distroless-static

which is before the commit I mentioned (https://github.com/vectordotdev/vector/commit/6dd9d65c85368200434ef68ae76430e8f7dbe64e)

and

jszwedkodd/vector:b8fb1e388-distroless-static

Which is the commit I think is likely to have caused the regression you are seeing.

Would you be able to try both of those?

**Comment by wtchangdm** [comment]
@jszwedko I've tested these two images:

On the left is `b8fb1e388-distroless-static`, another is `vector:b8fb1e388-distroless-static`.

![img](https://i.imgur.com/kr05XNr.png)

**Comment by wtchangdm** [comment]
Side note: I compared the rendered result of the official and mine, there are not many differences in the workload. 

And I took some time to run the official chart on this cluster with only service account changes and the `customConfig` you posted above.

The left is 0.20.1 and another is 0.21.0. Same result. No new information here though.
![img](https://i.imgur.com/ojsLiKz.png)

**Comment by jszwedko** [comment]
Hi @wtchangdm !

Thanks for trying that! Can you confirm which is which? In your comment you list the same image tag twice. I'm assuming that the one on the right is actually `6dd9d65-distroless-static` though.

Assuming that I'm correct, I built another image with an additional metric that I think might help figure out what's happening here. Would you mind trying to run `b8fb1e388-distroless-static` again (the baseline) followed by this new image tag: `829e39b34-distroless-static`? Additionally, for the second image tag, could you capture a plot of the added metric: `vector_component_received_events_count`? This is a histogram metric. It looks like you are using Prometheus/Grafana so I think you could plot the p50 and max?

**Comment by wtchangdm** [comment]
> I'm assuming that the one on the right is actually 6dd9d65-distroless-static though.

@jszwedko Yes, you are right. The left is `b8fb1e388-distroless-static` and the right is `6dd9d65-distroless-static`. Sorry for didn't double-check after copy-pasting. 

I will do the rest of tests and provide the metrics as well.

P.S. A not important side note: I tried the official chart instead with the images we discussed here and still got the same results. I was wondering if the in-house chart is the issue here. No new information here though.

**Comment by wtchangdm** [comment]
@jszwedko The result is here:

One the left is `b8fb1e388-distroless-static`, and the right is `829e39b34-distroless-static`.
![img](https://i.imgur.com/jTy3sU3.png)

The query used for p50 `vector_component_received_events_count_bucket`:

```
histogram_quantile(0.5, rate(vector_component_received_events_count_bucket{component_id!~"internal_metrics|prometheus_exporter"}[1m]))
```

max `vector_component_received_events_count_bucket` (seems stuck at `128`):

```
histogram_quantile(1, rate(vector_component_received_events_count_bucket{component_id!~"internal_metrics|prometheus_exporter"}[1m]))
```

Hope I got that right.

**Comment by jszwedko** [comment]
Thanks for trying that @wtchangdm ! That does give us some more info. I'm on vacation next week, but I'll plan to pick up this thread the week after.
